### PR TITLE
P2321R2 zip

### LINKED ISSUE #4987: [2021-10 LWG Motion 16] P2321R2 zip
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2321r2.html

### PR DESCRIPTION
**Pull request by jensmaurer**
- Fix incorrect template-heads for tuple constructors taking a pair.
- Rephrased "expression within noexcept" to "exception specification".

Fixes #4987
Fixes cplusplus/papers#1003

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -7430,3 +7476,2391 @@
...
+
+\pnum
+\indexlibrarymember{zip}{views}%
+The name \tcode{views::zip} denotes
+a customization point object\iref{customization.point.object}.
+Given a pack of subexpressions \tcode{Es...},
+the expression \tcode{views::zip(Es...)} is expression-equivalent to
+\begin{itemize}
+\item
+\tcode{\exposid{decay-copy}(views::empty<tuple<>>)}
```
Comment (cpplearner) [inline source/ranges.tex:7497]: Should this now use `auto(views::empty<tuple<>>)`?

File: source/utilities.tex [hunk 2]
```diff
@@ -1336,35 +1498,58 @@
...
+    template<class... UTypes>
+      constexpr const tuple& operator=(const tuple<UTypes...>&) const;
     template<class... UTypes>
       constexpr tuple& operator=(tuple<UTypes...>&&);
+    template<class... UTypes>
+      constexpr const tuple& operator=(tuple<UTypes...>&&) const;
 
     template<class U1, class U2>
-      constexpr tuple& operator=(const pair<U1, U2>&);          // only if \tcode{sizeof...(Types) == 2}
+      constexpr tuple& operator=(const pair<U1, U2>&);                  // only if \tcode{sizeof...(Types) == 2}
```
Comment (timsong-cpp) [inline source/utilities.tex:1542]: These comments seems to be too far to the right - the `==2` part is overflowing into a new line.

File: source/utilities.tex [hunk 3]
```diff
@@ -1517,130 +1702,81 @@
...
 \begin{itemize}
-\item \tcode{sizeof...(Types)} is 2,
-\item \tcode{is_constructible_v<$\tcode{T}_0$, U1>} is \tcode{true}, and
-\item \tcode{is_constructible_v<$\tcode{T}_1$, U2>} is \tcode{true}.
+\item
+\tcode{sizeof...(Types)} is 2,
+\item
+\tcode{is_constructible_v<$\tcode{T}_0$, decltype(get<0>(FWD(u)))>} is \tcode{true}, and
+\item
+\tcode{is_constructible_v<$\tcode{T}_1$, decltype(get<1>(FWD(u)))>} is \tcode{true}.
```
Comment (timsong-cpp) [inline source/utilities.tex:1766]:
```suggestion
\tcode{is_constructible_v<$\tcode{T}_0$, decltype(get<0>(\exposid{FWD}(u)))>} is \tcode{true}, and
\item
\tcode{is_constructible_v<$\tcode{T}_1$, decltype(get<1>(\exposid{FWD}(u)))>} is \tcode{true}.
```

File: source/ranges.tex [hunk 4]
```diff
@@ -7430,3 +7476,2389 @@
...
+\effects
+Equivalent to: \tcode{return y.\exposid{end_} - x.\exposid{current_}.back();}
+\end{itemdescr}
+
+\rSec2[range.adjacent.transform]{Adjacent transform view}
+
+\rSec3[range.adjacent.transform.overview]{Overview}
+
+\pnum
+\tcode{adjacent_transform_view} takes an invocable object and
```
Comment (JohelEGP) [inline source/ranges.tex:9363]: Missing index entry.

File: source/ranges.tex [hunk 5]
```diff
@@ -7430,3 +7476,2389 @@
...
+a \libconcept{view} and produces a \libconcept{view}
+whose $M^\textrm{th}$ element is the result of applying the invocable object
+to the $M^\textrm{th}$ through $(M + N - 1)^\textrm{th}$ elements
+of the original view.
+If the original view has fewer than $N$ elements, the resulting view is empty.
+
+\pnum
+The name \tcode{views::adjacent_transform<N>} denotes
```
Comment (JohelEGP) [inline source/ranges.tex:9371]: Missing index entry.

File: source/ranges.tex [hunk 6]
```diff
@@ -7430,3 +7476,2389 @@
...
+a range adaptor object\iref{range.adaptor.object}.
+Given subexpressions \tcode{E} and \tcode{F} and
+a constant expression \tcode{N}:
```
Comment (JohelEGP) [inline source/ranges.tex:9374]:
```suggestion
Given subexpressions \tcode{E} and \tcode{F}, and
a constant expression \tcode{N}:
```

File: source/ranges.tex [hunk 7]
```diff
@@ -7430,3 +7476,2389 @@
...
+\begin{itemize}
+\item
+If \tcode{N} is equal to 0,
```
Comment (JohelEGP) [inline source/ranges.tex:9377]:
```suggestion
If \tcode{N} is equal to \tcode{0},
```
Like above.

File: source/ranges.tex [hunk 8]
```diff
@@ -7430,3 +7476,2389 @@
...
+\end{example}
+
+\rSec3[range.adjacent.transform.view]{Class template \tcode{adjacent_transform_view}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{forward_range}@ V, @\libconcept{copy_constructible}@ F, size_t N>
+   requires @\libconcept{view}@<V> && (N > 0) && is_object_v<F> &&
+            @\libconcept{regular_invocable}@<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...> &&
+            @\exposid{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```
Comment (JohelEGP) [inline source/ranges.tex:9407]:
```suggestion
            @\exposconcept{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```

File: source/ranges.tex [hunk 9]
```diff
@@ -7430,3 +7476,2389 @@
...
+    adjacent_view<V, N> @\exposid{inner_}@;                 // \expos
+
+    using @\exposid{InnerView}@ = adjacent_view<V, N>;      // \expos
+    template<bool Const>
+    using @\exposid{inner-iterator}@ = iterator_t<@\exposid{maybe-const}@<Const, @\exposid{InnerView}@>>;  // \expos
+    template<bool Const>
+    using @\exposid{inner-sentinel}@ = sentinel_t<@\exposid{maybe-const}@<Const, @\exposid{InnerView}@>>;  // \expos
+
+    template<bool> class @\exposid{iterator}@;              // \expos
+    template<bool> class @\exposid{sentinel}@;              // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:9419]: Missing their cross reference.

File: source/ranges.tex [hunk 10]
```diff
@@ -7430,3 +7476,2389 @@
...
+\end{itemdescr}
+
+\rSec3[range.adjacent.transform.iterator]{Class template \tcode{adjacent_transform_view::\exposid{iterator}}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{forward_range}@ V, @\libconcept{copy_constructible}@ F, size_t N>
+    requires @\libconcept{view}@<V> && (N > 0) && is_object_v<F> &&
+             @\libconcept{regular_invocable}@<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...> &&
+             @\exposid{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```
Comment (JohelEGP) [inline source/ranges.tex:9483]:
```suggestion
             @\exposconcept{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```

File: source/ranges.tex [hunk 11]
```diff
@@ -7430,3 +7476,2389 @@
...
+    using value_type =
+      remove_cvref_t<invoke_result_t<@\exposid{maybe-const}@<Const, F>&,
+                                     @\exposid{REPEAT}@(range_reference_t<@\exposid{Base}@>, N)...>>;
+    using difference_type = range_difference_t<@\exposid{Base}@>;
+
+    @\exposid{iterator}@() = default;
+    constexpr @\exposid{iterator}@(@\exposid{iterator}@<!Const> i)
+      requires Const && @\libconcept{convertible_to}@<@\exposid{inner-iterator}@<false>, @\exposid{inner-iterator}@<Const>>;
+
+    constexpr decltype(auto) operator*() const noexcept(see below);
```
Comment (JohelEGP) [inline source/ranges.tex:9505]:
```suggestion
    constexpr decltype(auto) operator*() const noexcept(@\seebelow@);
```

File: source/ranges.tex [hunk 12]
```diff
@@ -7430,3 +7476,2389 @@
...
+Equivalent to:
+\begin{codeblock}
+return apply([&](const auto&... iters) -> decltype(auto) {
+  return invoke(*@\exposid{parent_}@->@\exposid{fun_}@, *iters...);
+}, @\exposid{inner_}@.@\exposid{current_}@);
+\end{codeblock}
+
+\pnum
+\remarks
+Let \tcode{Is} be the pack \tcode{0, 1, ..., (N-1)}.
```
Comment (JohelEGP) [inline source/ranges.tex:9614]:
```suggestion
Let \tcode{Is} be the pack \tcode{0, 1, \ldots, (N-1)}.
```

File: source/ranges.tex [hunk 13]
```diff
@@ -7430,3 +7476,2389 @@
 \effects
 Equivalent to: \tcode{return x.\exposid{end_} - y.\exposid{current_};}
 \end{itemdescr}
+
+\rSec2[range.zip]{Zip view}
+
+\rSec3[range.zip.overview]{Overview}
+
+\indexlibraryglobal{zip_view}%
```
Comment (timsong-cpp) [inline source/ranges.tex:7484]:
```suggestion
\pnum
\indexlibraryglobal{zip_view}%
```

File: source/ranges.tex [hunk 14]
```diff
@@ -7430,3 +7476,2389 @@
...
+\end{itemdescr}
+
+\rSec3[range.adjacent.transform.sentinel]{Class template \tcode{adjacent_transform_view::\exposid{sentinel}}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{forward_range}@ V, @\libconcept{copy_constructible}@ F, size_t N>
+    requires @\libconcept{view}@<V> && (N > 0) && is_object_v<F> &&
+             @\libconcept{regular_invocable}@<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...> &&
+             @\exposid{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```
Comment (JohelEGP) [inline source/ranges.tex:9788]:
```suggestion
             @\exposconcept{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
```

File: source/ranges.tex [hunk 15]
```diff
@@ -7430,3 +7476,2389 @@
...
+\end{itemdescr}
+
+\begin{itemdecl}
+template<bool OtherConst>
+  requires @\libconcept{sized_sentinel_for}@<@\exposid{inner-sentinel}@<Const>, @\exposid{inner-iterator}@<OtherConst>>
+friend constexpr range_difference_t<@\exposid{maybe-const}@<OtherConst, @\exposid{InnerView}@>>
+  operator-(const @\exposid{iterator}@<OtherConst>& x, const @\exposid{sentinel}@& y);
+
+template<bool OtherConst>
+  requires @\libconcept{sized_sentinel_for}@<inner-sentinel<Const>, @\exposid{inner-iterator}@<OtherConst>>
```
Comment (JohelEGP) [inline source/ranges.tex:9855]:
```suggestion
  requires @\libconcept{sized_sentinel_for}@<@\exposid{inner-iterator}@<Const>, @\exposid{inner-iterator}@<OtherConst>>
```

File: source/ranges.tex [hunk 16]
```diff
@@ -7430,3 +7476,2389 @@
...
+              @\libconcept{regular_invocable}@<F&, range_reference_t<Views>...> &&
+              @\exposid{can-reference}@<invoke_result_t<F&, range_reference_t<Views>...>>
+  template<bool Const>
+  class zip_transform_view<F, Views...>::@\exposid{iterator}@ {
+    using @\exposid{Parent}@ = @\exposid{maybe-const}@<Const, zip_transform_view>;      // \expos
+    using @\exposid{Base}@ = @\exposid{maybe-const}@<Const, @\exposid{InnerView}@>;                 // \expos
+    @\exposid{Parent}@* @\exposid{parent_}@ = nullptr;                                  // \expos
+    @\exposid{ziperator}@<Const> @\exposid{inner_}@;                                    // \expos
+
+    constexpr @\exposid{iterator}@(@\exposid{Parent}@& parent, @\exposid{ziperator}@<Const> inner); // exposition only
```
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/ranges.tex:8340]: Or put it on its own line, as done elsewhere.

File: source/ranges.tex [hunk 17]
```diff
@@ -7430,3 +7476,2391 @@
...
+      (@\libconcept{bidirectional_range}@<@\exposid{maybe-const}@<Const, Views>> && ...);
+  template<bool Const, class... Views>
+    concept @\defexposconcept{all-forward}@ =                            // \expos
+      (@\libconcept{forward_range}@<@\exposid{maybe-const}@<Const, Views>> && ...);
+
+  template<@\libconcept{input_range}@... Views>
+    requires (@\libconcept{view}@<Views> && ...) && (sizeof...(Views) > 0)
+  template<bool Const>
+  class zip_view<Views...>::@\exposid{iterator}@ {
+    @\exposid{tuple-or-pair}@<iterator_t<@\exposid{maybe-const}@<Const, Views>>...> @\exposid{current_}@;       // \expos
```
[the remark this answers is not in the document]
Comment (timsong-cpp) [reply source/ranges.tex:7663]: The default constructors of `tuple` and `pair` value-initialize their elements.

File: source/ranges.tex [hunk 18]
```diff
@@ -7430,3 +7476,2389 @@
...
+\end{itemdecl}
+
+\begin{itemdescr}
+\effects
+Equivalent to:
+\begin{codeblock}
+using ST = decltype(ranges::size(@\exposid{base_}@));
+using CT = common_type_t<ST, size_t>;
+auto sz = static_cast<CT>(ranges::size(@\exposid{base_}@));
+sz -= std::min<CT>(sz, N - 1);
```
[the remark this answers is not in the document]
Comment (timsong-cpp) [reply source/ranges.tex:8845]: The `std::` is intentional. We don't want to pick up `ranges::min`.

If `N > sz` then `N -1 >= sz`, and so `std::min<CT>(sz, N - 1)` is `sz`? I fail to see the problem.

File: source/ranges.tex [hunk 19]
```diff
@@ -7430,3 +7476,2420 @@
...
+\begin{itemdescr}
+\pnum
+\effects
+Initializes \exposid{current_} with \tcode{std::move(current)}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr @\exposid{iterator}@(@\exposid{iterator}@<!Const> i)
+  requires Const &&
+    (@\libconcept{convertible_to}@<iterator_t<Views>, iterator_t<@\exposid{maybe-const}@<Const, Views>>> && ...);
```
Comment (burblebee) [inline source/ranges.tex:7773]: line up with "Const" for consistency

File: source/ranges.tex [hunk 20]
```diff
@@ -7430,3 +7476,2420 @@
...
+  template<bool Const>
+  class zip_view<Views...>::@\exposid{sentinel}@ {
+    @\exposid{tuple-or-pair}@<sentinel_t<@\exposid{maybe-const}@<Const, Views>>...> @\exposid{end_}@;              // \expos
+    constexpr explicit @\exposid{sentinel}@(@\exposid{tuple-or-pair}@<sentinel_t<@\exposid{maybe-const}@<Const, Views>>...> end);
+                                                                                // \expos
+  public:
+    @\exposid{sentinel}@() = default;
+    constexpr @\exposid{sentinel}@(@\exposid{sentinel}@<!Const> i)
+      requires Const &&
+        (@\libconcept{convertible_to}@<sentinel_t<Views>, sentinel_t<@\exposid{maybe-const}@<Const, Views>>> && ...);
```
Comment (burblebee) [inline source/ranges.tex:8105]: line up with "Const" for consistency

File: source/ranges.tex [hunk 21]
```diff
@@ -7430,3 +7476,2420 @@
...
+\begin{itemdescr}
+\pnum
+\effects
+Initializes \exposid{end_} with \tcode{end}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr @\exposid{sentinel}@(@\exposid{sentinel}@<!Const> i)
+  requires Const &&
+    (@\libconcept{convertible_to}@<sentinel_t<Views>, sentinel_t<@\exposid{maybe-const}@<Const, Views>>> && ...);
```
Comment (burblebee) [inline source/ranges.tex:8140]: line up with "Const" for consistency?

**Review by burblebee — CHANGES_REQUESTED** [review]
great job on the edits, hardly any real errors! amazing!

File: source/ranges.tex [hunk 22]
```diff
@@ -7430,3 +7476,2420 @@
...
+
+\pnum
+\remarks
+The exception specification is equivalent to
+\begin{codeblock}
+(noexcept(ranges::iter_move(declval<const iterator_t<@\exposid{maybe-const}@<Const,
+                                                            Views>>&>())) && ...) &&
+(is_nothrow_move_constructible_v<range_rvalue_reference_t<@\exposid{maybe-const}@<Const,
+                                                                      Views>>> && ...)
+\end{codeblock}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex:8062]: We have this for "Effects: Equivalent to"
https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#writing-effects-in-a-function-description
I note that an expression is not preceded by a colon, and that seems like a good rule.

File: source/utilities.tex [hunk 23]
```diff
@@ -1517,130 +1703,81 @@
...
-
-\begin{itemdescr}
 \pnum
 \constraints
 \begin{itemize}
-\item \tcode{sizeof...(Types)} is 2,
-\item \tcode{is_constructible_v<$\tcode{T}_0$, U1>} is \tcode{true}, and
-\item \tcode{is_constructible_v<$\tcode{T}_1$, U2>} is \tcode{true}.
+\item
+\tcode{sizeof...(Types)} is 2,
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/utilities.tex:1784]: We seems to say "is" and then a small number without \tcode in the vicinity, so I think I'll leave it as-is for local consistency.

File: source/ranges.tex [hunk 24]
```diff
@@ -7430,3 +7476,2420 @@
...
+\rSec3[range.adjacent.overview]{Overview}
+
+\pnum
+\indexlibraryglobal{adjacent_view}%
+\tcode{adjacent_view} takes a \libconcept{view} and
+produces a \libconcept{view} whose $M^\text{th}$ element is
+a tuple of references to
+the $M^\text{th}$ through $(M + N - 1)^\text{th}$ elements of
+the original view.
+If the original view has fewer than $N$ elements, the resulting view is empty.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex:8904]: See @jwakely's comment above. Not changing.

File: source/utilities.tex [hunk 25]
```diff
@@ -1517,130 +1702,81 @@
...
-\item
-\tcode{is_constructible_v<$\tcode{T}_i$, const $\tcode{U}_i$\&>} is \tcode{true} for all $i$, and
-
-\item
-either
-\tcode{sizeof...(Types)} is not 1, or
-(when \tcode{Types...} expands to \tcode{T} and \tcode{UTypes...} expands to \tcode{U})
-\tcode{is_convertible_v<const tuple<U>\&, T>}, \tcode{is_constructible_v<T, const tuple<U>\&>}, and \tcode{is_same_v<T, U>} are all \tcode{false}.
-\end{itemize}
+Let \tcode{I} be the pack \tcode{0, 1, ..., (sizeof...(Types) - 1)}.\newline
```
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/utilities.tex:1735]: We have a fair amount of precedent for (ab)using code dots like this, so I don't want to solve this right now. We can take a wider look later if we wish.

File: source/ranges.tex [hunk 26]
```diff
@@ -7430,3 +7476,2391 @@
...
+
+    constexpr auto end() requires (!@\libconcept{simple-view}@<V>) {
+      if constexpr (@\libconcept{common_range}@<V>) {
+        return @\exposid{iterator}@<false>(@\exposid{as-sentinel}@{}, ranges::begin(@\exposid{base_}@), ranges::end(@\exposid{base_}@));
+      } else {
+        return @\exposid{sentinel}@<false>(ranges::end(@\exposid{base_}@));
+      }
+    }
+
+    constexpr auto end() requires @\libconcept{range}@<const V> {
```
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/ranges.tex]: Fixed!

File: source/ranges.tex [hunk 27]
```diff
@@ -7430,3 +7476,2389 @@
...
+\begin{itemize}
+\item
+If
+\begin{codeblock}
+invoke_result_t<@\exposid{maybe-const}@<Const, F>&, range_reference_t<@\exposid{maybe-const}@<Const, Views>>...>
+\end{codeblock}
+is not an lvalue reference,
+\tcode{iterator_category} denotes \tcode{input_iterator_tag}.
+\item
+Otherwise, let \tcode{Cs...} denote the pack of types
```
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/ranges.tex:8562]: Fixed.

File: source/ranges.tex (hunk 6 above)
Comment (tkoeppe) [reply source/ranges.tex:9539]: The status quo seems fine, and consistent with a previous use of "given...".

File: source/ranges.tex [hunk 28]
```diff
@@ -7430,3 +7476,2420 @@
...
+\item
+Otherwise, if \exposid{Base} models \libconcept{bidirectional_range},
+then \tcode{iterator_concept} denotes \tcode{bidirectional_iterator_tag}.
+\item
+Otherwise, \tcode{iterator_concept} denotes \tcode{forward_iterator_tag}.
+\end{itemize}
+
+\pnum
+If the invocation of any non-const member function of \exposid{iterator}
+exits via an exception, the \exposid{iterator} acquires a singular value.
```
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/ranges.tex:9094]: The first occurrence refers to the type ("member function of the type"), whereas the second refers to an instance, so the definite article seems useful here. By contrast, "iterator acquires a singular value" seems hard to parse.

"What value": a singular value (that's a well-defined notion).

"How": I don't think it's too hard to imagine, i.e. presumably via a try/catch block.
